src/mobu/events.py

```diff
@@ -0,0 +1,100 @@
+"""App metrics events."""
+
+from datetime import timedelta
+from typing import override
+
+from safir.dependencies.metrics import EventDependency, EventMaker
+from safir.metrics import EventManager, EventPayload
+
+
+class EventBase(EventPayload):
+    """Attributes on every mobu event."""
+
+    flock: str | None
+    business: str
+    username: str
+
+
+class NotebookBase(EventBase):
+    """Attributes for all notebook-related events."""
+
+    notebook: str
+    repo: str
+    repo_ref: str
+
+
+class NotebookExecution(NotebookBase):
+    """Reported after a notebook is finished executing."""
+
+    duration: timedelta | None
+    success: bool
+
+
+class NotebookCellExecution(NotebookBase):
+    """Reported after a notebook cell is finished executing."""
+
+    duration: timedelta | None
+    cell_id: str
+    success: bool
+    contents: str
```

Comment:
Agreed that a commit hash would be more useful than storing the actual contents. If we wanted to query by that hash, either in a where clause or in a group by, it would have to be an Influx tag, and that could contribute to high cardinality, but it's worth a shot.

I don't think a hash is useful if we're not going to be able to make it a tag. I imagine that the cell execution timings would mainly be useful if we see a big change in the notebook execution timings, and then we would want to see durations/errors by hash on both the notebook and cell execution events.

Reply:
I think I'm missing something: can't we dump the timings and the hashes by notebook and cell and see exactly when the timing degraded? In practice it's probably going to be rather visible on a graph against time, and then we just look at the data point before and after the change and see what hashes they used and that gives us the commit. I can see how this would be a bit easier if it were a tag, but it doesn't seem that important to make it a tag.